ipa-core/src/helpers/transport/in_memory/handlers.rs

```diff
@@ -0,0 +1,102 @@
+use std::{collections::HashSet, future::Future};
+
+use crate::{
+    helpers::{
+        query::{PrepareQuery, QueryConfig},
+        transport::in_memory::{routing::Addr, transport::Error, InMemoryTransport},
+        HelperIdentity, RouteId, Transport, TransportCallbacks, TransportIdentity,
+    },
+    protocol::QueryId,
+    sharding::ShardIndex,
+};
+
+///
+pub trait RequestHandler<I: TransportIdentity>: Send {
+    fn handle(
+        &mut self,
+        transport: InMemoryTransport<I>,
+        addr: Addr<I>,
+    ) -> impl Future<Output = Result<(), Error<I>>> + Send;
+}
+
+/// Helper trait to bind in-memory request handlers to transport identity.
+pub trait IdentityHandlerExt: TransportIdentity {
+    type Handler: RequestHandler<Self>;
+}
```

Comment:
Do these new traits really need to be different for in-memory vs. network transport?

One strategy to simplify might be to move `TransportCallbacks` up a level. Instead of having five special purpose `Box<dyn>` handlers in `TransportCallbacks`, use `Box<dyn Handler>` (or `Box<dyn TransportMessageHandler>` for a more specific name) as the type of the `callbacks: I::Handler` argument.

The relationship amongst these traits also seems complicated to me. It might be possible to define something like:

```
trait Transport {
   type Identity;
   type Handler;
}

impl Transport for ShardTransport {
   type Identity = ShardIndex;
   type Handler = ...;
}

impl Transport for MpcTransport {
   type Identity = HelperIdentity;
   type Handler = HelperRequestHandler;
}
```

(This doesn't by itself do anything about the duplication between net and memory transports, however.)

Reply:
I was hoping that using dynamic dispatch might make it easier to make pieces of this the same across transport types (helper, and shard, and ideally HTTP as well).

Maybe something like `fn listen` takes `Option<&dyn Handler>` as the callbacks argument? (The HTTP transport doesn't have `fn listen`, but it does use `TransportCallbacks`.) The shard transport can omit the handler entirely (because it only needs to handle records, at least for now), and for all transport types they no longer need to enumerate the management request types ("prepare query", "receive query", etc.), which feels like something that transports don't need to know about, and that shouldn't need to behave differently across the different transport types.

Reply:
I removed `IdentityHandlerExt` in favour of `ListenerSetup` - makes it a little less confusing imo.
I am not sure I fully understand your proposal, will sync up with you offline

Reply:
I took a look at `ListenerSetup` and I like it 👍.

Reply:
We discussed it offline and agreed that `TransportCallbacks` needs to go away. A better interface for handlers would be similar to [this one](https://docs.rs/iron/latest/iron/middleware/trait.Handler.html). 

It is still worth to encapsulate the delivery method inside `Transport` interface, so messages delivered via HTTP or potentially CF workers channel have the same structure when handled by application logic. We may have a good abstraction for it already - `Addr` struct used inside in-memory network. 

So our handler interface may look like this

```rust
trait Handler<I: TransportIdentity> {
   fn handle(&self, _: &Addr<I>) -> Result<Response>
}
```

HTTP transport can parse HTTP requests into `Addr` struct and application will have a unified logic to process requests based on `Addr:RouteId` field

Reply:
This issue is actually blocking now because `prepare_inputs` handler will require all transports to properly initialize the gateway. Naive approach would make MPC transport (that handles callbacks) to depend on Shard transport implementation (just to pass it to Gateway) which would make things way more complicated.

I am thinking that I would implement something like this

```mermaid
flowchart LR
  HelperApp --> Transports
  HelperApp --> QueryProcessor
  HelperApp --> Handler
  Handler --> Transports
  Handler --> QueryProcessor
```